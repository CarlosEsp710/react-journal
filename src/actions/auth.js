import {
  createUserWithEmailAndPassword,
  getAuth,
  signInWithEmailAndPassword,
  signInWithPopup,
  signOut,
  updateProfile,
} from "firebase/auth";
import Swal from "sweetalert2";

import { finishLoading, startLoading } from "./ui";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";

const auth = getAuth();

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((error) => {
        dispatch(finishLoading());
        Swal.fire("Error", error.message, "error");
      });
  };
};

export const startRegisterEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((error) => {
        dispatch(finishLoading());
        Swal.fire("Error", error.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => {});
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
  };
};

export const logout = () => ({ type: types.logout });
