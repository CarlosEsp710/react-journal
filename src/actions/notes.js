import {
  addDoc,
  collection,
  deleteDoc,
  doc,
  updateDoc,
} from "firebase/firestore";
import Swal from "sweetalert2";

import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const doc = await addDoc(
      collection(db, `${uid}`, "journal/notes"),
      newNote
    );

    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    const activeNote = getState().notes.active;

    if (!activeNote.url) {
      delete activeNote.url;
    }

    const noteToFirestore = { ...activeNote };
    delete noteToFirestore.id;
    const noteRef = doc(db, `${uid}/journal/notes/${activeNote.id}`);

    await updateDoc(noteRef, noteToFirestore);

    dispatch(refreshNote(activeNote.id, noteToFirestore));

    Swal.fire("Saved", "Your note has been saved", "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const activeNote = getState().notes.active;

    Swal.fire({
      title: "Please Wait!",
      html: "Data uploading...",
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;

    Swal.close();

    dispatch(startSaveNote(activeNote));
  };
};

export const startDeletingNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    const activeNote = getState().notes.active;

    const noteRef = doc(db, `${uid}/journal/notes/${activeNote.id}`);
    await deleteDoc(noteRef);

    dispatch(deleteNote(activeNote.id));
  };
};

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const logoutCleaning = () => ({
  type: types.notesLogoutCleaning,
});
