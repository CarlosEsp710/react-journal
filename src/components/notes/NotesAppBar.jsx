import React from "react";

export const NotesAppBar = () => {
  return (
    <div className="notes__app-bar">
      <span>28 de agosto del 2021</span>
      <div>
        <button className="btn cursor">Picture</button>
        <button className="btn cursor">Save</button>
      </div>
    </div>
  );
};
