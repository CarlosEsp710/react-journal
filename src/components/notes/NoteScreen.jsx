import React, { useEffect, useRef } from "react";

import { useDispatch, useSelector } from "react-redux";

import { activeNote, startDeletingNote } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const [form, handleInputChange, reset] = useForm(note);
  const { body, title } = form;

  const activeId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(form.id, { ...form }));
  }, [form, dispatch]);

  const handleDelete = () => dispatch(startDeletingNote());

  return (
    <div className="notes__main-content animate__animated animate__fadeIn animate__faster">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          name="title"
          placeholder="Some awesome title"
          className="notes__title-input"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          name="body"
          placeholder="What happened today?"
          className="notes__textarea"
          value={body}
          onChange={handleInputChange}
        ></textarea>
        {note.url && (
          <div className="notes__image">
            <img src={note.url} alt={note.url} />
          </div>
        )}
      </div>

      <button className="btn btn-danger" onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};
