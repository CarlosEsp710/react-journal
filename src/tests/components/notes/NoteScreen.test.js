import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from "../../../actions/notes";

jest.mock("../../../actions/notes", () => ({
  activeNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {
    uid: "123",
    name: "John Doe",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    active: {
      id: "123",
      title: "",
      body: "",
    },
    notes: [],
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <NoteScreen />
    </MemoryRouter>
  </Provider>
);
describe("test <NoteScreen />", () => {
  test("should return <NoteScreen /> component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call activeNote", () => {
    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "Hola mundo",
      },
    });

    expect(activeNote).toHaveBeenCalled();
    expect(activeNote).toHaveBeenLastCalledWith("123", {
      id: "123",
      title: "Hola mundo",
      body: "",
    });
  });
});
