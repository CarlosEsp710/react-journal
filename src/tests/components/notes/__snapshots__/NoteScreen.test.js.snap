// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <NoteScreen /> should return <NoteScreen /> component 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__app-bar"
  >
    <span>
      28 de agosto del 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn cursor"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn cursor"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value=""
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value=""
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
